Async datastore

andrei-sa opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/eidolonsystems/beam/1)
<!-- Reviewable:end -->

kranar committed:
Added starter code for the AsyncDataStore.

Commit:
Implement AsyncDataStore

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,216 @@
+      m_flushPending(false),
+      m_currentDataStore(std::make_shared<ReserveDataStore>()),
+      m_flushedDataStore(m_currentDataStore) {}
+
```
Remove blank line.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,216 @@
+      dataStore.swap(m_currentDataStore);
+      m_flushPending = false;
+    }
+    m_dataStore->Store(dataStore->LoadAll());
```
The issue as I see it is that if an ```AsyncDataStore::Load``` happens from another thread while the current thread is executing line 208, then none of the data in ```dataStore``` will be returned. Can you correct this bug in ```AsyncDataStore```? I can then apply the fix to ```BufferedDataStore```.

andrei-sa replied:
I don't quite get why. Load reads data from `m_flushedDataStore` in addition to `m_dataStore`. `m_flushedDataStore` is normally bound to `m_currentDataStore` (they are 2 copies of the same shared pointers), but on line 205, `m_currentDataStore` gets temporarily detached from from `m_flushedDataStore` because it is replaced with a new local data store to allow for new writes during the flush process. `m_flushedDataStore` still points to the old data store, up to line 211, which is now being flushed using the `dataStore` handle. As a result, `Load` will retrieve all entries from `m_dataStore` and `dataStore`, removing possible duplicates. Am I missing something?

kranar replied:
Okay, then Load needs to load from m_dataStore to get all persisted data... m_flushedDataStore to get all data currently being flushed, and m_currentDataStore to get any data that has not been flushed before and is not currently being flushed, correct?

andrei-sa replied:
Okay, I think I am getting that now. If the flush process has started but not yet finished, new Stores occur followed by Loads, Loads won't see data from those new Stores as they are writing to `m_currentDataStore` not yet bound to `m_flushedDataStore`.
Reading from all three data store handles does not seem promising performance-wise because simultaneously reading from and writing to `m_currentDataStore` would require locking excessive locking.
I can, however, try to only trigger the heavy `m_currentDataStore` read only when there is a flush process in progress. That can be published using another flag, which is checked inside the critical section in `Load`. If the flush is happening at that time, all entries are loaded from `m_currentDataStore` while the critical section is still open. A new flush process can begin anywhere after the critical section ends, but it does not matter as only new Stores (called after Load was called) will possibly be unaccounted for.

andrei-sa replied:
Should we follow the approach described above, we should probably use a shared lock as loaders may get heavily interlocked during `m_currentDataStore` loads.

kranar replied:
As I mention in my other comment, feel free to take some liberties to implement a solution as you see fit, given the time constraints and implementation requirements. You can always submit an initial implementation now that is reasonably fast, and then work on further optimizations later.

Commit:
Remove unnecessary blank line

andrei-sa commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,215 @@
+    } else {
+      matches = buffer->Load(query);
+    }
+    if(static_cast<int>(matches.size()) < query.GetSnapshotLimit().GetSize()) {
```
I have another concern about this one. You mentioned that usually the data added are new, but it is not strictly guaranteed, nor does the API restrict users to that.
Having said that, older data may be present in the buffer than those available in the main data store. This logic, however, relying on data in the buffer being always newer.
As an example of unexpected behavior, consider `BufferedDataStore` with a buffer size of 100. My main data store contains all the new data, but I have received a data feed with 100 old entries. I insert store them in my `BufferedDataStore`; those elements now fully occupy the buffer. I then attempt to load 100 newest entries from `BufferedDataStore`, but instead of getting the actual head of the principal data feed I get those values from the lateral feed.

It seems to me that I have to load `SnapshotLimit` items from every data store involved and then select a total of `SnapshotLimit` newest/oldest according to the query.

andrei-sa replied:
Another option is to load newest/oldest entries one-by-one and merge on the fly, but it may be less efficient depending on the data store used. Would also have to pre-load `m_currentDataStore` anyway to avoid locking the mutex for the entire merge process.

kranar replied:
Sure, implement a solution based on what you think will best accomplish the goal. Every DataStore implementation is required to be synchronized. The BufferedDataStore is available as a guide to help you implement the AsyncDataStore, but it's not intended to be the actual solution, so don't worry too much about it. It might contain bugs, unexpected behavior, or other issues, none of which should be present in the AsyncDataStore.

Commit:
Account for out-of-order Store and Store during Flush

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+
+namespace Beam::Queries {
+namespace Details {
+  template<typename T>
```
Eliminate the use of subtype polymorphism. It isn't providing any benefit since all types are known at compile time and hence can be resolved at compile time.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+      Matches(Iterator begin, Iterator end)
+        : m_begin(std::move(begin)),
+          m_end(std::move(end)),
+          m_cur(m_begin + 1),
```
Undefined behavior when ```m_begin == m_end```.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+        : m_begin(std::move(begin)),
+          m_end(std::move(end)),
+          m_cur(m_begin + 1),
+          m_nextValue(*m_begin) {}
```
Undefined behavior when ```m_begin == m_end```.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+        if(m_cur == m_end) {
+          m_cur = m_begin;
+        } else {
+          m_nextValue = *m_cur;
```
Try to avoid a copy here, perform a move if possible.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+      Iterator m_begin;
+      Iterator m_end;
+      Iterator m_cur;
+      Type& m_nextValue;
```
As a coding style, we don't use references as fields but rather pointers, so change ```Type&``` to ```Type*```.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+  };
+
+  template<typename I>
+  std::unique_ptr<Matches<I>> MakeMatches(I begin, I end) {
```
Do not dynamically allocate memory for this.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+  }
+
+  template<typename T>
+  std::tuple_element_t<0, T> MergeMatches(T matches,
```
Pass in the arguments individually rather than constructing a ```tuple```. If you still need to use a tuple then construct one from within ```MergeMatches```, but preferably avoid the need for a tuple altogether.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+  std::tuple_element_t<0, T> MergeMatches(T matches,
+      const SnapshotLimit& limit) {
+    using Type = typename std::tuple_element_t<0, T>::value_type;
+    auto data = std::list<std::unique_ptr<BaseMatches<Type>>>();
```
Consider a ```deque``` instead of a ```list```.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+        }
+      };
+      (testInclude(parts), ...);
+    }, std::move(matches));
```
You use move semantics on line 86, but never take advantage of move semantics from line 75 to line 85. See if you can take advantage of move semantics from line 75-85, otherwise do not move ```matches```.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+    auto result = std::vector<Type>();
+    for(auto i = 0; !data.empty() && i < limit.GetSize(); ++i) {
+      auto partIter = data.begin();
+      auto comparator = [](auto& lhs, auto& rhs) {
```
Prefer ```const auto&``` instead of ```auto&``` for polymorphic lambda parameters.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+        return SequenceComparator()(lhs->PeekNext(), rhs->PeekNext());
+      };
+      if(limit.GetType() == SnapshotLimit::Type::HEAD) {
+        partIter = std::min_element(data.begin(), data.end(), comparator);
```
Is this neccessary? One property of DataStore's is that they always return values sorted in ascending order by sequence number, so this may not be needed.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+      auto lock = boost::lock_guard(m_mutex);
+      buffer = m_flushedDataStore;
+      if(m_flushInProgress) {
+        latestMatches = m_currentDataStore->Load(query);
```
Do not perform a ```Load``` within a lock. Hoist the ```Load``` out of the critical section.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+    }
+    auto bufferedMatches = m_flushedDataStore->Load(query);
+    auto dataStoreMatches = m_dataStore->Load(query);
+    auto consolidatedMatches = std::make_tuple(std::move(latestMatches),
```
As mentioned earlier, pass the arguments individually to ```MergeMatches``` rather than constructing a ```tuple``` and then moving it in.

kranar commented on Beam/Include/Beam/Queries/AsyncDataStore.hpp:
```diff
@@ -0,0 +1,295 @@
+    m_tasks.Push(
+      [=] {
+      Flush();
+    });
```
Fix indentation.

Commit:
Remove runtime polymorphism from MergeMatches implementation

Commit:
Merge remote-tracking branch 'origin/master' into async_datastore

Commit:
Pre-allocate result array in MergeMatches

Commit:
Remove extra Flush from AsyncDataStore::Shutdown

Commit:
Add async tests using TestDataStore

Commit:
Reduce the number of scheduled flush tasks by using FlushStatus

Commit:
Remove locks from Store

kranar committed:
Removed dynamic memory allocation.
Removed several redundant moves by inlining expressions directly.
Unfolded tuple used to merge matches.
Removed undefined behavior when using std::min.
Eliminated data race when performing a Store.

kranar committed:
Fixed a bug when merging matches together.

kranar committed:
Coding style.

kranar committed:
Merge branch 'master' into async_datastore

kranar committed:
WIP, refactoring the MergeMatches function.

kranar committed:
Identified and applies code reductions.
Manually unrolled the test to determine which match has the next element (previously was using min_element).

kranar committed:
Reduced the amount of state needed to merge matches together.

kranar committed:
Eliminated copy's/move's when merging matches.

kranar committed:
Added template deduction guide for AsyncDataStore.
Removed the TestFlushFrequency unit test, since that is no longer a goal of the AsyncDataStore.
Added a free function to automatically do the work needed to open a TestDataStore.
